File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

int diameter (TreeNode* root){

    if (root == NULL) return NULL;

    int leftheight = diameter(root-> left);
    int rightheight = diameter(root->right);

    return 1+ max(leftheight, rightheight);
}

    int diameterOfBinaryTree(TreeNode* root) {
        if (root == NULL) return NULL;

        int leftDiam = diameterOfBinaryTree(root-> left);
        int rightDiam = diameterOfBinaryTree(root-> right);
        int current = diameter(root-> left) + diameter(root-> right);

        return max((current), max(leftDiam, rightDiam));



       
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    boolean ans = true;
    public boolean isBalanced(TreeNode root) {
        // return dfsHeight(root) != -1;
        dfsHeight(root);
        return ans;


    }
    private int dfsHeight(TreeNode node) {
        if (node==null) return 0;

    int lefth = dfsHeight(node.left);
    // if (lefth == -1) return -1;

    int righth = dfsHeight(node.right);
    // if (righth == -1) return -1;

    if (Math.abs(lefth-righth) >1){
    ans = false;
    }

    return Math.max(lefth,righth)+1;
    }
}
